torch/csrc/autograd/python_variable.cpp

```diff
@@ -1941,24 +1942,55 @@ static PyObject* THPVariable_NewWithVar(
     PyTypeObject* type,
     Variable _var,
     c10::impl::PyInterpreterStatus status) {
-  // This function overwrite the Tensor's pyobj field without extra checks
-  // Make sure it is not set otherwise we would leak memory
-  auto mb_obj = _var.unsafeGetTensorImpl()->pyobj_slot()->check_pyobj(
-      self_interpreter.get());
-  TORCH_CHECK(
-      !mb_obj.has_value() || !mb_obj.value(),
-      "Creating a new Tensor subclass ",
-      type->tp_name,
-      " but the raw Tensor object is already associated to a python object ",
-      "of type ",
-      mb_obj.value()->ob_type->tp_name);
-
   // Make sure that the reinterpret into a THPVariable* will be valid
   TORCH_CHECK(
       PyType_IsSubtype(type, &THPVariableType),
       "Creating a Tensor subclass from a class ",
       "that does not inherit from Tensor is not possible. Make sure your class inherits from Tensor.");
 
+  // This function overwrite the Tensor's pyobj field without extra checks
+  // Make sure it is not set otherwise we would leak memory
+  auto mb_obj = _var.unsafeGetTensorImpl()->pyobj_slot()->check_pyobj(
+      self_interpreter.get());
+
+  // Under some circumstances, we may attempt to create a new Python
+  // object for a variable that already has a Python object.  The most common
+  // situation this can occur is if you have a TorchDispatchMode active that
+  // is returning a subclass from lift_fresh (which is invoked to
+  // appropriately "wrap" a constant tensor into whatever ambient modes are
+  // active.)
+  //
+  // In general, it is impossible to handle this case compositionally.
+  // Suppose you have a user call ATensor(...) when a mode is active that
+  // is transforming all ops to output BTensor, where ATensor and BTensor
+  // are completely unrelated subclasses and there is no way to compose them.
+  // There is no way to satisfy the user request here; we must error.
+  //
+  // However, a more common case is a user just called torch.Tensor(...),
```

Comment:
Small question about what we want `t2 = torch.Tensor(t1)` to be?
Here you seem to imply that `t2 is t1`. Is that what we want? Or are we saying that `t2` is a view of `t1` and is desugared into `t2 = t1.lift()` (and the lift is done above before this new function is called)?

Reply:
Definitely not the former, and that the comment made you think that means I need to reword the comment.

But the latter is not right either, because we do not call lift() in the torch.Tensor(t1) case where t1 is a Tensor. `legacy_tensor_generic_ctor_new` says that in this case, we call t1.alias() for this case. The lift case occurs when you say `torch.Tensor([1, 2, 3])`, at which point we create a raw tensor representing [1, 2, 3] (with all modes disabled), and then we call `lift_fresh` to apply modes to it.

What IS true is that in either of these cases, the alias/lift_fresh may return BTensor instead of Tensor. If the user had just requested torch.Tensor at the call site, we will allow it (returning that BTensor directly), instead of raising an error.